src/Microsoft.AspNet.Hosting.Server.Abstractions/IHttpApplication.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.AspNet.Http.Features;
+
+namespace Microsoft.AspNet.Hosting.Server
+{
+    /// <summary>
+    /// Represents an HttpApplication.
+    /// </summary>
+    public interface IHttpApplication<THttpContext>
```

Comment:
Much prefer this design. I'm wondering if we should have BeginRequest/ProcessRequest/EndRequest. If we did that i'm not sure how to couple creation of the context object. Would it be on the side or interleaved? Maybe I'm just being anal :smile: 

```C#
var context = app.BeginRequest(features);

try
{
     await app.ProcessRequest(context);
}
finally
{
    app.EndRequest(context);
}
```